Various improvements

stephank opened the pull request:
Now uses pyjwt, explains how to install in a virtualenv, and has a small config file.

I'd very much like to reduce logic even further, if this is to be a proper example eventually. E.g., the 'fetch config and keys' dance is something that seems like a perfect candidate for hiding behind an abstraction.

But I suppose that kind of stuff is OpenID Connect specific bits shining through, which is unavoidable, unless we want to pull in a large dependency like pyoidc.

It'd be great if we could have an even shorter example using off-the-shelve libs, and avoid having to provide our own lib for every language and platform out there.

Other than that, I'm curious what other changes we'd like to see in this repo.

stephank commented:
This now tests and depends on these two portier-broker PRs:
- https://github.com/portier/portier-broker/pull/13
- https://github.com/portier/portier-broker/pull/16

In addition, I've added comments to `server.py`.

stephank commented:
Thinking about it, indexing nonces by email is perhaps a bad example, because the final authenticated email may not match the login hint. We could:
- Index by IP instead, which is also more obviously 'examplish'.
- Use cookies, which is more work. (I wonder if set-cookie in the 303 response works?)

stephank commented:
Added a bit more polish, and now indexing by IP, which is a small step. We can always add the larger step of cookies, if we want.

djc commented on requirements.txt:
```diff
@@ -1 +1,7 @@
-cryptography
+cffi==1.8.3
```
Is this list auto-generated by something? In general, I don't like making dependencies stricter than necessary. So I would expect just adding `PyJWT` here.

stephank replied:
This is `pip freeze` output. Shouldn't we at least pin major versions? Apparently, we can do this: https://www.python.org/dev/peps/pep-0440/#compatible-release

djc commented on server.py:
```diff
@@ -42,8 +43,8 @@ def login(env):
     body = env['wsgi.input'].read(int(env['CONTENT_LENGTH']))
-    jwt = parse.parse_qs(body)[b'id_token'][0].decode('ascii')
-    result = get_verified_email(jwt)
+    token = parse.parse_qs(body)[b'id_token'][0].decode('ascii')
```
Would you mind splitting this hunk out into a separate commit?

djc commented on server.py:
```diff
@@ -68,15 +69,7 @@ def b64dec(s):
-    # [x] exp > (now) > iat, with some margin
-    # [-] sub is a valid email address
-
+def get_verified_email(token):
```
This could also belong in that other commit.

djc commented on server.py:
```diff
@@ -102,38 +95,13 @@ def get_verified_email(jwt):
-    )))
     try:
-        verifier.verify()
+        payload = jwt.decode(token, pub_key,
```
Did you check that PyJWT actually does at least as much validation as we did before this?

stephank replied:
Yes, PyJWT documents these: http://pyjwt.readthedocs.io/en/latest/usage.html#registered-claim-names

djc commented on README.rst:
````diff
@@ -5,15 +5,19 @@ This is a minimal Python implementation of a Portier Relying Party. It is
 
 ```bash
-# Optional: create a virtualenv
+# Optional: Create a virtualenv.
````
Please put these style tweaks in a separate commit.

Also, like in the broker PR, most of these comments feel overly explicit/obvious to me.

djc commented on server.py:
```diff
@@ -70,10 +63,9 @@ def b64dec(s):
-        META['PORTIER_ORIGIN'],
-        '/.well-known/openid-configuration',
-    )))
+    rsp = request.urlopen(
```
Here, too, I'd prefer the formatting changes (which I otherwise like for having one less line) in a separate commit, so that this one ist just about the config thingy.

djc commented on README.rst:
````diff
@@ -4,22 +4,22 @@ Portier Relying Party demo code
-# Start the server.
-./server.py
-```
+.. code-block:: bash
````
Please squash this into the commit introducing this block.

djc commented on server.py:
```diff
@@ -32,8 +32,31 @@ def index(env):
+    body = env['wsgi.input'].read(int(env['CONTENT_LENGTH']))
+    email = parse.parse_qs(body)[b'email'][0].decode('ascii')
+
+    auth_url = '%(portier_origin)s/auth?%(qs)s' % {
```
The way you're using formatting strings here seems kind of weird. The `%(name)s` style seems useful mainly if (a) you already have a dict containing the variable's names, or (b) if there are lots of variables being inserted. Neither of these is the case here, so it seems to me that just using `%s` would probably be more readable, and would prevent the silly repetition of `'portier_origin': META['portier_origin']`. `qs` can just be a local variable, after which you can do the formatting in place in the `return` statement, I think.

djc commented on server.py:
```diff
@@ -32,8 +32,31 @@ def index(env):
+            'scope': 'openid email',
+            'response_type': 'id_token',
+            'client_id': META['rp_origin'],
+            'redirect_uri': '%(rp_origin)s/verify' % {
```
Here, you could probably just do `% META`, instead of restating the obvious?

djc commented on server.py:
```diff
@@ -15,29 +15,44 @@
 
 DIR = os.path.dirname(os.path.abspath(__file__))
 
+# Load the configuration.
```
These again feel very fine-grained and fairly obvious repetition of the actual code. Can you tighten up by removing some and/or updating them with more why/context?

djc commented on server.py:
```diff
@@ -86,30 +114,61 @@ def b64dec(s):
     n = int.from_bytes(b64dec(key['n']), 'big')
     pub_key = rsa.RSAPublicNumbers(e, n).public_key(default_backend())
 
+    # Verify the token. We let the pyjwt do a lot of work for us, but you
```
Adding this after you just removed something similar a few commits back feels pretty silly. Maybe do it in a smarter way, morphing the previous comment into what you want here? (Still feels more useful to provide high-level stuff in the docstring, as it makes it more accessible than just having a comment.)

djc commented on server.py:
```diff
@@ -118,24 +177,48 @@ def get_verified_email(token):
+    # Take the email address from the 'subject' claim in the payload, and
+    # verify the format of the email address is what we expect.
     sub = payload['sub']
     if not re.match('.+@.+', sub):  # <-- TODO: Use a proper parser.
```
I think we should probably get rid of the TODO here. At most, we can check for the rules used in the `emailaddress` Rust crate we're using for the broker, where I think it also ends up using a totally simple algorithm for this in the end, because it's most web-compatible and trying to parse email addresses exactly right is just not worth it in terms of complexity.

stephank replied:
`emailaddress` does exactly as much is this right now, so will just get rid of the TODO.

djc commented on server.py:
```diff
@@ -4,20 +4,21 @@
 from time import time
 from wsgiref import simple_server
-import os, mimetypes, html
+import os, re, json, jwt, binascii, mimetypes, html
```
Please don't reformat in the same commit.

djc commented on server.py:
```diff
@@ -35,6 +35,13 @@ def template(tpl, status=200, **vars):
     return 200, headers, src.encode('utf-8')
 
 
+def parse_post_body(env):
```
Nice. This makes me wonder if we can normalize the return value a bit more, so that the users don't have to do as much song-and-dance to get their value out. E.g., decode keys to unicode strings, take last value of list by default, and decode the resulting value as well.

djc commented on server.py:
```diff
@@ -15,7 +15,7 @@
 META = json.load(open(os.path.join(DIR, 'config.json')))
 
-# Our nonce storage: a dict indexed by email. Normally, you'd store nonces with
+# Our nonce storage: a dict indexed by IP. Normally, you'd store nonces with
```
I think this is the wrong direction. I think it's a strong invariant for the broker to always return the same email address as requested by the RP, anyway; if normalization is needed, this should be handled in the broker, hiding it from the RP.

Since normalization is probably not even consistent across hosts, I think this is a requirement for the broker, so it makes sense to rely on that.

djc commented on server.py:
```diff
@@ -174,14 +174,11 @@ def get_verified_email(token, nonce):
     #  - Check that expiration time is not in the past
     #    (The `exp` claim in the payload).
     #
-    #  - Check that the 'not before' time is not in the future
```
This should also probably get squashed into previous iterations of this comment.

stephank commented:
@djc Think I've addressed all feedback.

Note that we still need https://github.com/portier/portier-broker/pull/16 merged before this, but that one should also be ready to merge!

djc commented on server.py:
```diff
@@ -127,6 +137,12 @@ def get_verified_email(token):
         return {'error': 'Invalid email: %s' % sub}
 
+    # Make sure the nonce cannot be used on further attempts.
+    nonce = NONCES[sub]
```
I'm missing code that will handle the case where the email address is not in `NONCES`, I think?

stephank replied:
I've fixed this, and found it actually make the code shorter in this case. But in several other places, we also don't check `payload`.

djc commented:
Great, thanks!
